ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncActions.java

```diff
@@ -0,0 +1,131 @@
+
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class FastSyncActions<C> {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private final SynchronizerConfiguration syncConfig;
+  private final ProtocolSchedule<C> protocolSchedule;
+  private final ProtocolContext<C> protocolContext;
+  private final EthContext ethContext;
+  private final LabelledMetric<OperationTimer> ethTasksTimer;
+
+  public FastSyncActions(
+      final SynchronizerConfiguration syncConfig,
+      final ProtocolSchedule<C> protocolSchedule,
+      final ProtocolContext<C> protocolContext,
+      final EthContext ethContext,
+      final LabelledMetric<OperationTimer> ethTasksTimer) {
+    this.syncConfig = syncConfig;
+    this.protocolSchedule = protocolSchedule;
+    this.protocolContext = protocolContext;
+    this.ethContext = ethContext;
+    this.ethTasksTimer = ethTasksTimer;
+  }
+
+  public CompletableFuture<Void> waitForSuitablePeers() {
+    final WaitForPeersTask waitForPeersTask =
+        WaitForPeersTask.create(
+            ethContext, syncConfig.getFastSyncMinimumPeerCount(), ethTasksTimer);
+
+    final EthScheduler scheduler = ethContext.getScheduler();
+    final CompletableFuture<Void> result = new CompletableFuture<>();
+    scheduler
+        .timeout(waitForPeersTask, syncConfig.getFastSyncMaximumPeerWaitTime())
+        .handle(
+            (waitResult, error) -> {
+              if (ExceptionUtils.rootCause(error) instanceof TimeoutException) {
+                if (ethContext.getEthPeers().bestPeer().isPresent()) {
+                  LOG.warn(
+                      "Fast sync timed out before minimum peer count was reached. Continuing with reduced peers.");
+                  result.complete(null);
+                } else {
+                  waitForAnyPeer()
+                      .thenAccept(result::complete)
+                      .exceptionally(
+                          taskError -> {
+                            result.completeExceptionally(error);
+                            return null;
+                          });
+                }
+              } else if (error != null) {
+                LOG.error("Failed to find peers for fast sync", error);
+                result.completeExceptionally(error);
+              } else {
+                result.complete(null);
+              }
+              return null;
+            });
+
+    return result;
+  }
+
+  private CompletableFuture<Void> waitForAnyPeer() {
+    LOG.warn(
+        "Maximum wait time for fast sync reached but no peers available. Continuing to wait for any available peer.");
+    final WaitForPeerTask waitForPeerTask = WaitForPeerTask.create(ethContext, ethTasksTimer);
+    return ethContext.getScheduler().scheduleSyncWorkerTask(waitForPeerTask::run);
```

Comment:
If we can't find any peers then there is literally nothing we can do but wait.  If we added a timeout, when it's reached we'd simply have to go back to waiting for a peer again, so why not just keep waiting?  We can't even fall back to full sync because it needs a peer to sync from as well.

It went to the worker pool for consistency but it doesn't actually need to.  Changed to call directly.

Reply:
Mostly I'm just thinking about how debugging would go in this case.  It might look like the synchronizer has hung if it just stops printing any logs.  If we did add a timeout, I would just go back and start a new `WaitForPeerTask` when the timeout triggered, but at least it's more obvious that it's still trying to sync.  Not a huge deal though.

Reply:
Yeah that's a good point.  I've changed it to repeatedly wait for peers so it prints the "Waiting for 1 peers" message repeatedly while it's waiting.